### PR TITLE
[telemetry] Add log rotation support (#7352)

### PR DESCRIPTION
**Pull request by h0cheung**
**Description:**
 Add log rotation support. The logs of otelcol itself will be rotated by lumberjack.

**Link to tracking Issue:**
#7352 

**Testing:**
Creating logger with or without rotation, logging with rotation and logging with an old big file are added in unit tests.

I did some simple testing on Linux, macOS, and Windows.

### CODE REVIEW

**Comment by h0cheung** [comment]
remove stale

**Comment by h0cheung** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

remove stale

**Comment by h0cheung** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

remove stale

**Comment by h0cheung** [comment]
Remove stale

File: config/configrotate/configrotate.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	MaxDays int `mapstructure:"max_days"`
+
+	// MaxBackups is the maximum number of old log files to retain. The default
+	// is to 100 files.
+	MaxBackups int `mapstructure:"max_backups"`
+
+	// LocalTime determines if the time used for formatting the timestamps in
+	// backup files is the computer's local time.  The default is to use UTC
+	// time.
+	LocalTime bool `mapstructure:"localtime"`
```
Comment (atoulme) [inline config/configrotate/configrotate.go:35]: Is the default false then?

File: config/configrotate/configrotate.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"io"
+	"os"
+
+	"gopkg.in/natefinch/lumberjack.v2"
+)
+
+type Config struct {
+	// Enabled controls whether or not rotate logs
```
Comment (atoulme) [inline config/configrotate/configrotate.go:14]: Is this true or false by default?

File: config/configrotate/configrotate.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	Enabled bool `mapstructure:"enable"`
```
Comment (atoulme) [inline config/configrotate/configrotate.go:15]:
```suggestion
	Enabled bool `mapstructure:"enabled"`
```

File: config/configrotate/configrotate.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	// MaxMegabytes is the maximum size in megabytes of the file before it gets
+	// rotated. It defaults to 100 megabytes.
+	MaxMegabytes int `mapstructure:"max_megabytes"`
```
Comment (atoulme) [inline config/configrotate/configrotate.go:19]:
```suggestion
	MaxMb int `mapstructure:"max_mb"`
```

File: config/configrotate/configrotate.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	// MaxDays is the maximum number of days to retain old log files based on the
+	// timestamp encoded in their filename.  Note that a day is defined as 24
+	// hours and may not exactly correspond to calendar days due to daylight
+	// savings, leap seconds, etc. The default is not to remove old log files
+	// based on age.
+	MaxDays int `mapstructure:"max_days"`
```
Comment (atoulme) [inline config/configrotate/configrotate.go:26]: So what is the default, 0?

File: config/configrotate/configrotate.go (hunk 4 above)
Comment (atoulme) [reply config/configrotate/configrotate.go:19]: Please add tests to check behavior if set to negative values. Same for max days.

File: config/configrotate/configrotate.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	// time.
+	LocalTime bool `mapstructure:"localtime"`
+}
+
+func (cfg *Config) NewWriter(filename string) (io.WriteCloser, error) {
+	if !cfg.Enabled {
+		// #nosec G302 G304 -- filename is a trusted safe path, and should allow to be read by other users
+		return os.OpenFile(filename, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+	}
+	return cfg.newLumberjackWriter(filename), nil
```
Comment (atoulme) [inline config/configrotate/configrotate.go:43]: this makes us lose the ability to issue an error on start if there are any config issues, any chance we can be alerted early if lumberjack will error out?

File: config/configrotate/configrotate.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (cfg *Config) newLumberjackWriter(filename string) io.WriteCloser {
+	return &lumberjack.Logger{
+		Filename:   filename,
```
Comment (atoulme) [inline config/configrotate/configrotate.go:48]: Can we test the behavior if the filename is invalid?

File: service/telemetry/config.go [hunk 8]
```diff
@@ -83,6 +84,19 @@ type LogsConfig struct {
...
+	// Example:
+	//
+	//     rotation:
+	//         max_megabytes: 100
+	//         max_days: 30
+	//         max_backups: 100
+	//         localtime: false
+	//
+	// By default, max size is 100MB before rotation. Max number of backups is 100,
+	// and no limit for days. UTC time will be used.
```
Comment (atoulme) [inline service/telemetry/config.go:98]: Is this already documented in the config struct? I'd avoid duplication here and remove those 2 lines.

Comment (atoulme) [reply service/telemetry/config.go:98]: also by default on shipping, I'd make sure enabled is off by default.

File: service/telemetry/telemetry.go [hunk 9]
```diff
@@ -91,3 +115,63 @@ func toSamplingConfig(sc *LogsSamplingConfig) *zap.SamplingConfig {
...
+// explanation: https://github.com/natefinch/lumberjack/pull/47#issuecomment-322502210
+type nopSyncSink struct {
+	io.WriteCloser
+}
+
+func (w nopSyncSink) Sync() error {
+	return nil
+}
+
+func setRotatinURL(paths []string, rotationSchema string) ([]string, error) {
```
Comment (atoulme) [inline service/telemetry/telemetry.go:140]:
```suggestion
func setRotatingURL(paths []string, rotationSchema string) ([]string, error) {
```

File: service/telemetry/telemetry.go [hunk 10]
```diff
@@ -91,3 +115,63 @@ func toSamplingConfig(sc *LogsSamplingConfig) *zap.SamplingConfig {
...
+				return nil, fmt.Errorf("ports not allowed with file URLs: got %v", u)
+			}
+			if hn := u.Hostname(); hn != "" && hn != "localhost" {
+				return nil, fmt.Errorf("file URLs must leave host empty or use localhost: got %v", u)
+			}
+
+			res = append(res, rotationSchema+":?path="+url.QueryEscape(u.Path))
+			continue
+		}
+		res = append(res, p)
```
Comment (atoulme) [inline service/telemetry/telemetry.go:174]: what happens if the scheme is http?

Comment (h0cheung) [reply service/telemetry/telemetry.go:174]: Zap allows us to register schemata. By default, only `file` is registered. Unregistered schemata will cause errors.

File: config/configrotate/configrotate.go (hunk 5 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:26]: Yes.

https://github.com/open-telemetry/opentelemetry-collector/blob/86f33b34b85681d316cb32fe0df25d8321cb6476/otelcol/unmarshaler.go#L52-L58

As there is no ServiceFactory. Default settings of service are in the unmarshaller.

File: config/configrotate/configrotate.go (hunk 1 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:35]: Yes.

https://github.com/open-telemetry/opentelemetry-collector/blob/86f33b34b85681d316cb32fe0df25d8321cb6476/otelcol/unmarshaler.go#L52-L58

As there is no ServiceFactory. Default settings of service are in the unmarshaller.

File: service/telemetry/config.go (hunk 8 above)
Comment (h0cheung) [reply service/telemetry/config.go:98]: I'll update these. Thanks

File: config/configrotate/configrotate.go (hunk 6 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:43]: Yeah, I'll do it

File: config/configrotate/configrotate.go (hunk 4 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:19]: The config was copied from [fileexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/fileexporter). I think we'd better use the same config for similar functions.

File: config/configrotate/configrotate.go (hunk 1 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:36]: https://github.com/open-telemetry/opentelemetry-collector/blob/ef274535cca9198797c26269dd1464ef60291d11/config/configrotate/configrotate.go#L39-L47

A default config func has been added.

File: config/configrotate/configrotate.go (hunk 3 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go]: It has been fixed

File: config/configrotate/configrotate.go (hunk 6 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:64]: This has been done in the latest commit.

File: config/configrotate/configrotate.go (hunk 7 above)
Comment (h0cheung) [reply config/configrotate/configrotate.go:69]: When writing logs, it will try to create parent directories. If failed, it will report errors (written to `error_output_paths`, `stderr` by default).
I've added checks to make sure the filename is valid before creating it.

File: service/telemetry/config.go (hunk 8 above)
Comment (h0cheung) [reply service/telemetry/config.go]: Updated in latest commit.

File: service/telemetry/telemetry.go (hunk 9 above)
Comment (h0cheung) [reply service/telemetry/telemetry.go]: Fixed in latest commit

**Comment by h0cheung** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

remove stale

**Comment by h0cheung** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

remove stale

**Comment by h0cheung** [comment]
remove stale

**Comment by kago-dk** [comment]
Did the log rotation ever make it into the code?

**Comment by djaglowski** [comment]
> Did the log rotation ever make it into the code?

It did not, but it should in some form. Looks like #7425 still tracks it. cc: @BinaryFissionGames.

**Comment by li-zeyuan** [comment]
Hi @djaglowski 
I think log rotation is a very important feature. Is there a plan to merge this PR? Or does this PR need to be perfected?

**Comment by djaglowski** [comment]
I agree @li-zeyuan. From what I can tell this timed out after feedback was implemented, but not rereivewed. @h0cheung are you willing to reopen this? @atoulme will you revisit your comments to confirm if they were addressed?

**Comment by h0cheung** [comment]
> I agree @li-zeyuan.
> 
> @open-telemetry/collector-maintainers, any concerns in principle with adding log rotation?
> 
> From what I can tell this timed out after feedback was implemented, but not re-reivewed. @h0cheung are you willing to reopen this? @atoulme will you revisit your comments to confirm if they were addressed?

Yes. I'd like to continue the work if we can reopen this.

**Comment by li-zeyuan** [comment]
> > I agree @li-zeyuan.
> > @open-telemetry/collector-maintainers, any concerns in principle with adding log rotation?
> > From what I can tell this timed out after feedback was implemented, but not re-reivewed. @h0cheung are you willing to reopen this? @atoulme will you revisit your comments to confirm if they were addressed?
> 
> Yes. I'd like to continue the work if we can reopen this.

@h0cheung Hi brother, can you add [repository permission](https://github.com/h0cheung/opentelemetry-collector/tree/feature/rotation) to me? Coding together.

**Comment by h0cheung** [comment]
> > > I agree @li-zeyuan.
> > > @open-telemetry/collector-maintainers, any concerns in principle with adding log rotation?
> > > From what I can tell this timed out after feedback was implemented, but not re-reivewed. @h0cheung are you willing to reopen this? @atoulme will you revisit your comments to confirm if they were addressed?
> > 
> > 
> > Yes. I'd like to continue the work if we can reopen this.
> 
> @h0cheung Hi brother, can you add [repository permission](https://github.com/h0cheung/opentelemetry-collector/tree/feature/rotation) to me? Coding together.

Yes

**Comment by djaglowski** [comment]
@li-zeyuan, are you still planning to update this PR?

**Comment by li-zeyuan** [comment]
> @li-zeyuan, are you still planning to update this PR?

There will be no time to update the PR in the next 2 months, but I will keep an eye on this feature.

**Comment by djaglowski** [comment]
@h0cheung if you'd be willing to grant me access I can update the PR.

File: service/telemetry/telemetry.go [hunk 11]
```diff
@@ -147,6 +154,22 @@ func newLogger(cfg LogsConfig, options []zap.Option) (*zap.Logger, error) {
 		zapCfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
 	}
 
+	if cfg.Rotation != nil && cfg.Rotation.Enabled {
+		rotationSchema := "rotation-" + uuid.NewString()
```
Comment (djaglowski) [inline service/telemetry/telemetry.go:158]: Do we need randomness here? Why can't we just register a reasonably unique schema, e.g. `otelcol-rotation`

File: service/telemetry/logger.go [hunk 12]
```diff
@@ -51,3 +76,67 @@ func newSampledLogger(logger *zap.Logger, sc *LogsSamplingConfig) *zap.Logger {
...
+// explanation: https://github.com/natefinch/lumberjack/pull/47#issuecomment-322502210
+type nopSyncSink struct {
+	io.WriteCloser
+}
+
+func (w nopSyncSink) Sync() error {
+	return nil
+}
+
+func setRotationURL(paths []string, rotationSchema string) ([]string, error) {
```
Comment (djaglowski) [inline service/telemetry/logger.go:101]: I'm unclear on what we can and cannot assume about the inputs to this function, and what we are trying to do with them. Are the "paths" constrained to `stdout`, `stderr` and file paths?

File: service/telemetry/telemetry.go (hunk 11 above)
Comment (h0cheung) [reply service/telemetry/telemetry.go]: As there will never be several loggers with different rotation config, I think `otelcol-rotation` is OK.

File: service/telemetry/logger.go (hunk 12 above)
Comment (h0cheung) [reply service/telemetry/logger.go:101]: As explained in [the codes of zap](https://github.com/uber-go/zap/blob/bd0cf0447951b77aa98dcfc1ac19e6f58d3ee64f/writer.go#L50), by default the only schema `file` is supported, and missing schema will be considered as `file`.

So I only modify URLs without schema or with schema `file`. URLs with other schemas will be added as is. 
https://github.com/h0cheung/opentelemetry-collector/blob/233cbf14e9d2769c21d5226d00c2d956c8559ffb/service/telemetry/logger.go#L116-139

**Comment by bvsvas** [comment]
@djaglowski I do agree this most important feature. What is the final outcome on this PR?

**Comment by djaglowski** [comment]
@bvsvas, if I remember correctly there were some failing unit tests on windows. I do not understand the URL related logic in this PR very well, especially how it relates to windows.

At this point I think the best path forward would be to comment on https://github.com/open-telemetry/opentelemetry-collector/issues/7352 with detailed design considerations and make sure there is buy in from the maintainers.

**Comment by yyildirim01** [comment]
> @bvsvas, if I remember correctly there were some failing unit tests on windows. I do not understand the URL related logic in this PR very well, especially how it relates to windows.
> 
> At this point I think the best path forward would be to comment on #7352 with detailed design considerations and make sure there is buy in from the maintainers.

Hi,

There are some topics related to log rotation, but it seems there have been no improvements or changes yet.
These are the ones that I found: #10768 #7507 #7352 

I wanted to ask how we can proceed on this and which item we should prioritize for a solution.
https://github.com/h0cheung/opentelemetry-collector/blob/233cbf14e9d2769c21d5226d00c2d956c8559ffb/service/telemetry/logger.go#L116-L139

Thank you.
